### PR TITLE
Add appdash/nettrace package

### PR DESCRIPTION
**Pull request by tcolgate**
First draft at tracer for adding events for an underlying net.Conn
Still needs tests and documentation, but I'd like to know if you'd
be interested in it if I tidy it up.

I'm using this as part of a benchmarking client for an API, which I
realise is a bit of an odd use of appdash (it's all client side), but the
code was easy to write and it seemed like it could be generally useful.

Feedback is welcome.

### CODE REVIEW

**Comment by slimsag** [comment]
Sounds like a reasonably well-fitted feature for Appdash. I have a few requests if you have the time:

- Would you be able to rename it to `nettrace` instead of `netconntrace`? I think that matches `httptrace` and `sqltrace` sub-packages better.
- Perhaps add some tests to ensure it works and package docs.
- Long term we should perhaps have a small example for it's usage in `appdash/examples/cmd/netapp`, but that can be added later.

Thanks for the PR! :smile:

**Comment by tcolgate** [comment]
I've split the Dial and net.Conn out into seperate files, and renamed as requested.
I'll have a crack at writing some tests at the end of the week. (I've never got the hang of unit testing this kind of stuff so it'll be a good thing for me to do).
I've got code using this in an internal project, so putting together an eample app should be pretty easy.

**Comment by slimsag** [comment]
Hey @tcolgate -- great work! I really like this! If you want any help or feedback on writing the tests just let me know, I'd be glad to help.

I added a bunch of comments on the first file regarding `// comment` style -- but they're very minor (and I can adjust them after merging if you don't want to). They also apply to the `ConnWriteEvent` type (the other file).

If you could put together a example app I would be very happy! If not just let me know, and I'll see if I can write one soon to verify that this API works pretty well.

Looking forward to your progress on this and to merging it! :smile:

**Comment by tcolgate** [comment]
Sorry this took a while, not had much spare time.
Updated the comments, and removed the two NewConn(Read|Write)Event.
Still need a unit test.

**Comment by slimsag** [comment]
Apologies for the late review. The code LGTM.

If you don't have time to write a unit test (or don't want to) I can merge this as-is and add one myself -- just let me know!

We really appreciate the PR! :smile:

**Comment by tcolgate** [comment]
@slimsag cheers, I'm going to hve a crack at the unit test. It'll be good practice.

**Comment by tcolgate** [comment]
feel free to merge first though, if you want to make it available.It might be a while before the test is ready (not got a lot of spare time in the next week or so, I will try today, if I don't get it done, it may be another week)

**Comment by tcolgate** [comment]
@slimsag I thought I'd update this and have a crack at adding some tests. Nothing too mind blowing yet, but I'll try and work on it. Will add Close() tracing too I think. I don't actually need this code any more, but it's fun to work on.

File: nettrace/doc.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Package nettrace implement a net.Dial and net.Conn that trace connections
+// reads, ands writes. An example of hooking this into the http Transport is below
```
Comment (shurcooL) [inline nettrace/doc.go]: Minor typo here. "ands" should be "and". Also, based on the sentence at the bottom, I am guessing it's missing a comma.

"that trace connections reads, ands writes" -> "that trace connections, reads, and writes"

File: nettrace/doc.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+//			Proxy:               http.ProxyFromEnvironment,
+//			Dial:                traceDial,
+//			TLSHandshakeTimeout: 10 * time.Second,
+//		}
+//
+//		httpClient := &http.Client{
+//			Transport: &httptrace.Transport{Recorder: rec, SetName: true, Transport: netTraceTransport},
+//		}
+//
+// This will cause connections, read and writes made by the httpClient to generate events.
```
Comment (shurcooL) [inline nettrace/doc.go]: Maybe "read" should be "reads" for consistency with above. Again really minor.

**Comment by slimsag** [comment]
@tcolgate Awesome! I appreciate it! If you ever feel like not working on it let me know and I can take over it. I think this will make a really good example on how Appdash can work in many areas, not just in HTTP applications :)

**Comment by slimsag** [comment]
I'm going to close this for now because it is stale. Once someone (incl. me) is ready to work on it, feel free to re-open / comment.

FWIW I do think this is a good change that we should make long-term.

**Pull request closed without merging**